*** main.c.org	Wed Feb 13 18:14:14 2002
--- main.c	Wed Feb 13 18:01:44 2002
***************
*** 176,181 ****
--- 176,183 ----
  	int	pr_usesysctl;		/* true if we use sysctl, not kvm */
  } protox[] = {
  	{ -1,		-1,		1,	protopr,
+ 	  sctp_stats,	NULL,		"sctp",	IPPROTO_SCTP },
+ 	{ -1,		-1,		1,	protopr,
  	  tcp_stats,	NULL,		"tcp",	IPPROTO_TCP },
  	{ -1,		-1,		1,	protopr,
  	  udp_stats,	NULL,		"udp",	IPPROTO_UDP },
***************
*** 199,204 ****
--- 201,208 ----
  
  #ifdef INET6
  struct protox ip6protox[] = {
+ 	{ -1,		-1,		1,	protopr,
+ 	  sctp_stats,	NULL,		"sctp",	IPPROTO_SCTP },
  	{ -1,		-1,		1,	protopr,
  	  tcp_stats,	NULL,		"tcp",	IPPROTO_TCP },
  	{ -1,		-1,		1,	protopr,
